src/storage/s3.rs

```diff
@@ -76,44 +86,53 @@ pub(super) struct S3StorageTransaction<'a> {
 }
 
 impl<'a> StorageTransaction for S3StorageTransaction<'a> {
-    fn store_batch(&mut self, batch: &[Blob]) -> Result<(), Error> {
-        let mut attempts = 0;
-
-        loop {
-            let mut futures = FuturesUnordered::new();
-            for blob in batch {
-                futures.push(
-                    self.s3
-                        .client
-                        .put_object(PutObjectRequest {
-                            bucket: self.s3.bucket.to_string(),
-                            key: blob.path.clone(),
-                            body: Some(blob.content.clone().into()),
-                            content_type: Some(blob.mime.clone()),
-                            content_encoding: blob.compression.as_ref().map(|alg| alg.to_string()),
-                            ..Default::default()
-                        })
-                        .inspect(|_| {
-                            crate::web::metrics::UPLOADED_FILES_TOTAL.inc_by(1);
-                        }),
-                );
-            }
-            attempts += 1;
-
-            match S3_RUNTIME.lock().block_on(futures.map(drop).collect()) {
-                // this batch was successful, start another batch if there are still more files
-                Ok(_) => break,
-                Err(err) => {
-                    error!("failed to upload to s3: {:?}", err);
-                    // if a futures error occurs, retry the batch
-                    if attempts > 2 {
-                        panic!("failed to upload 3 times, exiting");
+    fn store_batch(&mut self, mut batch: Vec<Blob>) -> Result<(), Error> {
+        self.s3.runtime.handle().block_on(async {
+            // Attempt to upload the batch 3 times
+            for _ in 0..3 {
+                let mut futures = FuturesUnordered::new();
+                for blob in batch.drain(..) {
+                    futures.push(
+                        self.s3
+                            .client
+                            .put_object(PutObjectRequest {
+                                bucket: self.s3.bucket.to_string(),
+                                key: blob.path.clone(),
```

Comment:
Pre-existing, but can we avoid this clone here? It takes a [`ByteStream`](https://docs.rs/rusoto_core/0.45.0/rusoto_core/struct.ByteStream.html), which it looks like we can construct with only [`&mut blob.path`](https://docs.rs/futures-core/0.3.5/futures_core/stream/trait.Stream.html#impl-Stream-for-%26%27_%20mut%20S).

Reply:
Unfortunately that's an `impl Stream for &mut S where S: ?Sized + Stream + Unpin`, which an `&mut Vec<u8>` doesn't satisfy

Reply:
Ah gotcha, I misread it. This is fine then. I think I actually opened an issue for this a while ago: https://github.com/rusoto/rusoto/issues/234#issuecomment-632423950